source/code/source/playfab/HttpClient.cpp

```diff
@@ -0,0 +1,227 @@
+#include <stdafx.h>
+#include <playfab/HttpClient.h>
+#include <playfab/PlayFabSettings.h>
+
+namespace PlayFab
+{
+    // RAII wrapper around HCCallHandle
+    class HCHttpCall
+    {
+    public:
+        static void Perform(
+            const char* url,
+            const UnorderedMap<String, String>& headers,
+            const char* requestBody,
+            HttpClient::HttpCallback&& callback
+        );
+
+        virtual ~HCHttpCall() noexcept;
+
+    private:
+        HCHttpCall(HttpClient::HttpCallback&& callback);
+        HCHttpCall(const HCHttpCall& other) = delete;
+        HCHttpCall& operator=(HCHttpCall other) = delete;
+
+        HRESULT Initialize(
+            const char* url,
+            const UnorderedMap<String, String>& headers,
+            const char* requestBody
+        );
+        void CompleteRequest(HRESULT networkErrorCode);
+
+        static void CALLBACK HCPerformComplete(XAsyncBlock* async);
+
+        HCCallHandle m_callHandle{ nullptr };
+        HttpClient::HttpCallback m_callback;
+        HttpResult m_result;
+        XAsyncBlock m_asyncBlock{};
+    };
+
+    HttpClient::HttpClient(SharedPtr<PlayFabApiSettings> settings) :
+        m_settings(std::move(settings))
+    {
+    }
+
+    void HttpClient::MakePostRequest(
+        const char* path,
+        const UnorderedMap<String, String>& headers,
+        const char* requestBody,
+        HttpCallback&& callback
+    ) const
+    {
+        if (m_settings->titleId.empty())
+        {
+            throw new PlayFabException(PlayFabExceptionCode::TitleNotSet, "PlayFabSettings::staticSettings->titleId has not been set properly. It must not be empty.");
+        }

```

Comment:
avoid exceptions for errors.  prefer HRESULTS up to caller

Reply:
ya error handling not final.  Still not clear exactly how titleId will be set so this exception is a holdover from the old http stack.

---
In reply to: [596448616](https://github.com/jplafonta/XPlatCTemplate/pull/1#discussion_r596448616) [](ancestors = 596448616)